File: #015/3Sum.java
public class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> lists = new LinkedList<>();
        for (int i = 0; i < nums.length-2; i++) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j < nums.length-1; j++) {
                if (nums[i] + nums[j] > 0) break;
                if (j - i > 1 && nums[j] == nums[j - 1]) continue;
                for (int k = j + 1; k < nums.length; k++) {
                    if (nums[i] + nums[j] + nums[k] > 0) break;
                    if (nums[i] + nums[j] + nums[k] == 0) {
                        List<Integer> integers = Arrays.asList(nums[i], nums[j], nums[k]);
                        if (!lists.contains(integers)) {
                            lists.add(integers);
                        }
                        break;
                    }
                }
            }
        }
        return lists;
    }
}

File: #015/3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> list;
        for (int i = 0; i < nums.size(); i++) {
            int target = -nums[i];
            int front = i + 1;
            int back = nums.size() - 1;
            while (front < back) {
                int sum = nums[front] + nums[back];
                if (sum < target) front++;
                else if (sum > target) back--;
                else {
                    vector<int> tmp(3, 0);
                    tmp[0] = nums[i];
                    tmp[1] = nums[front];
                    tmp[2] = nums[back];
                    list.push_back(tmp);
                    while (front < back && nums[front] == tmp[1]) front++;
                    while (front < back && nums[back] == tmp[2]) back--;
                }
            }
            while (i + 1 < nums.size() && nums[i + 1] == nums[i])i++;
        }
        return list;
    }
};